Automate sites.yaml config

kevinmeijer97 opened the pull request:
We can automate the generation of the new sites.yaml.
By doing this we reduce the amount of config needed to configure the sites, only keeping a few settings in the sites array of `config/rapidez/statamic.php`.

Not sure yet if setting this during boot is the way to go, could also generate it through a command when setting up the site.
Let me know what you would prefer!

royduin commented on src/RapidezStatamicServiceProvider.php:
```diff
@@ -55,6 +55,31 @@ public function boot()
+        foreach ($stores as $store) {
+            $sites[$store['code']] = [
+                'name' => $store['name'] ?? $store['code'],
+                'locale' => '{{ config:rapidez.statamic.sites.' . $store['code'] . '.locale }}',
+                'lang' => '{{ config:rapidez.statamic.sites.' . $store['code'] . '.lang }}',
+                'url' => '{{ config:rapidez.statamic.sites.' . $store['code'] . '.url }}',
```
We can also get this from Magento right? The locales and url's are also configured there?

royduin commented on src/RapidezStatamicServiceProvider.php:
```diff
@@ -55,6 +55,31 @@ public function boot()
+    public function bootSites() : self
+    {
+        $sites = [];
+        $stores = Rapidez::getStores();
```
We shouldn't query anything directly from the Service Provider, as it's executed even from your console when you didn't have a database connection yet. But if we move it to a middleware it's only executed on browser requests; not when you run something in the CLI. But this generates the sites config file? Maybe a command?

BobWez98 commented on src/Commands/CreateSites.php:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        Site::setSites($sites);
+        Site::save();
```
Since we're getting this from the database anyway, can't we maybe just overwrite the `getSavedSites` function in `Statamic\Sites\Sites` to just return a cached array with the queries already done? Then we set the sites on the fly and we can get rid of the sites.yaml all together 🤔

kevinmeijer97 replied:
Yes i agree, i feel like this might be the best way to go.
I cached the sites array for now because of the setStore() function which is needed to get the correct site data.
After https://github.com/rapidez/core/pull/567 gets merged we can use getValue() function without having to do the setStore().

BobWez98 reviewed (commented):
This feels a bit manual, maybel we can add some magic to do it on the fly, but this is just food for thought.

kevinmeijer97 commented on src/Commands/CreateSites.php:
```diff
@@ -0,0 +1,46 @@
+<?php
```
Yes good catch, explained further in https://github.com/rapidez/statamic/pull/81#discussion_r1806395996

royduin commented on src/Extend/SitesLinkedToMagentoStores.php:
```diff
@@ -20,4 +22,38 @@ public function findByMageRunCode($code)
+                    'url' => $url,
+                    'attributes' => [
+                        'magento_store_id' => $store['store_id'],
+                        'group' => $store['website_code'] ?? ''
```
Previously the group was explained in the readme, with this PR that info will be removed. Maybe we should add some info on the alternate hreflang link tags in the readme?

kevinmeijer97 replied:
Added some info for this into the readme

royduin commented on src/Extend/SitesLinkedToMagentoStores.php:
```diff
@@ -20,4 +22,38 @@ public function findByMageRunCode($code)
+                ];
+            }
+
+            return $sites ?: $this->getFallbackConfig();
```
The method was changed recently: https://github.com/statamic/cms/pull/10470, do we require the versions where that's changed? https://github.com/statamic/cms/releases/tag/v5.16.0

royduin commented on src/RapidezStatamicServiceProvider.php:
```diff
@@ -33,7 +32,7 @@ class RapidezStatamicServiceProvider extends ServiceProvider
     public function register()
     {
         $this->app->extend(Sites::class, function () {
-            return new SitesLinkedToMagentoStores(config('statamic.sites'));
+            return new SitesLinkedToMagentoStores();
         });
```
This can be rewritten to a oneliner now?

kevinmeijer97 commented on config/rapidez/statamic.php:
```diff
@@ -80,13 +80,10 @@
 
     'sites' => [
```
Changed the readme for this!

kevinmeijer97 commented on src/Extend/SitesLinkedToMagentoStores.php:
```diff
@@ -20,4 +22,38 @@ public function findByMageRunCode($code)
+
+                $locale = $configModel::getCachedByPath('general/locale/code');
+                $lang = explode('_', $locale)[0] ?? '';
+                $url = $configModel::getCachedByPath('web/secure/base_url');
```
This happened to be some missing Rapidez config in the Magento repo. The .env of the Magento repo missed the Rapidez urls which led to the `web/secure/base_url` config being the Magento url instead of the Rapidez url.

kevinmeijer97 commented:
I think we should wait with merging so we can release it in bulk along with the Hybrid runway changes.

BobWez98 commented on src/RapidezStatamicServiceProvider.php:
```diff
@@ -32,9 +31,7 @@ class RapidezStatamicServiceProvider extends ServiceProvider
-        $this->app->extend(Sites::class, function () {
-            return new SitesLinkedToMagentoStores(config('statamic.sites'));
-        });
+        $this->app->extend(Sites::class, fn () => new SitesLinkedToMagentoStores());
```
Isn't `$this->app->extend(Sites::class, SitesLinkedToMagentoStores::class);` enough?

kevinmeijer97 replied:
No the second parameter had to be a closure, keeping it on this for now

royduin commented on README.md:
```diff
@@ -24,7 +24,9 @@ composer require rapidez/statamic
-It will mainly setup the [Eloquent driver](https://github.com/statamic/eloquent-driver) and publish the necessary vendor from the rapidez/statamic repo. 
+It will mainly setup the [Eloquent driver](https://github.com/statamic/eloquent-driver) and publish the necessary vendor from the rapidez/statamic repo.
+
+Before running the install script, make sure you have an existing User model as Statamic requires this. See: [Statamic docs](https://statamic.dev/tips/storing-users-in-a-database)
```
This doesn't explain how you should get you user model back: https://github.com/laravel/laravel/blob/11.x/app/Models/User.php, as we don't provide it from `rapidez/rapidez`: https://github.com/rapidez/rapidez/blob/040ada124619473f7d713aab537d82cea51f0026/.github/workflows/copy-laravel-source.yml#L45. You'll also need the migrations, etc.

kevinmeijer97 replied:
Changed the readme with a clearer explanation for how and what with the user model!

kevinmeijer97 commented:
Moved most of the configuration guide to the Rapidez docs (https://github.com/rapidez/docs/pull/56).

BobWez98 commented on config/rapidez/statamic.php:
```diff
@@ -80,13 +80,10 @@
-                'magento_store_id' => 1,
-                'group' => 'default',
                 'disabled' => false,
             ],
```
Do we still need this whole sites config?

kevinmeijer97 replied:
Changed this around, see https://github.com/rapidez/statamic/pull/81#issuecomment-2457255941

BobWez98 commented on src/Extend/SitesLinkedToMagentoStores.php:
```diff
@@ -20,4 +22,38 @@ public function findByMageRunCode($code)
+            $configModel = config('rapidez.models.config');
+
+            foreach ($stores as $store) {
+                if (config('rapidez.statamic.sites.' . $store['code'] . '.attributes.disabled')) {
```
Can't we safely say we always want this to be enabled? And otherwise maybe just have an array that includes the disabled stores?

kevinmeijer97 replied:
Changed this around, see https://github.com/rapidez/statamic/pull/81#issuecomment-2457255941

BobWez98 commented:
We can merge this in my opinion, what about yours @royduin

royduin commented:
We've a conflict but besides that I'm also curious about the answers on your latest questions. It's currently unclear where the `sites` in the config is for or when to use it.

kevinmeijer97 commented:
> We've a conflict but besides that I'm also curious about the answers on your latest questions. It's currently unclear where the `sites` in the config is for or when to use it.

Conflict has been fixed!
Also refactored the sites array and replaced it with a `disabled_sites` array, there is no need to have a sites array in the config anymore when the only thing configurable is the disabled option.
You can add the store_codes of the stores you don't want in Statamic to the `disabled_sites` array.

kevinmeijer97 commented on config/rapidez/statamic.php:
```diff
@@ -78,17 +78,7 @@
-    ]
+    'disabled_sites' => [
+        // '{site_handle}'
+    ],
```
Changed the docs for this: https://github.com/rapidez/docs/pull/62

kevinmeijer97 commented on src/Extend/SitesLinkedToMagentoStores.php:
```diff
@@ -22,6 +24,40 @@ public function findByMageRunCode($code)
+
+    protected function getSavedSites()
+    {
+        return Cache::rememberForever('statamic_sites', function () {
```
Changed the docs for this: https://github.com/rapidez/docs/pull/62
